### PR TITLE
Fix spans bug when scatter or client_desires_new_key creates a task

### PR DESCRIPTION
**Pull request by crusaderky**
Of the whole test suite, this use case was triggered (not always) exclusively by test_queue.py::test_race.

### CODE REVIEW

File: distributed/scheduler.py [hunk 1]
```diff
@@ -4448,13 +4450,15 @@ def update_graph(
...
         spans_ext: SpansExtension | None = self.extensions.get("spans")
         if spans_ext:
-            span_annotations = spans_ext.new_tasks(new_tasks)
-            if span_annotations:
-                resolved_annotations["span"] = span_annotations
-            else:
-                # Edge case where some tasks define a span, while earlier tasks in the
-                # same TaskGroup don't define any
-                resolved_annotations.pop("span", None)
+            # observe_tasks does not necessarily contain all runnable tasks;
```
Comment (hendrikmakait) [inline distributed/scheduler.py:4453]: Did you mean `new_tasks`?
```suggestion
            # new_tasks does not necessarily contain all runnable tasks;
```

File: distributed/tests/test_spans.py [hunk 2]
```diff
@@ -258,35 +327,103 @@ def update_graph(self, scheduler, *, tasks, annotations, **kwargs):
...
+    await wait_for_state("x-0", "released", s)
+    assert s.tasks["x-0"].group.span_id is None
+    x1 = c.submit(inc, 1, key="x-1")
+    assert await x1 == 2
+    assert s.tasks["x-0"].group.span_id is not None
+
+
+@gen_cluster(client=True)
+async def test_scatter_creates_ts(c, s, a, b):
+    """A TaskState object is created by scatter, and only later becomes runnable
```
Comment (hendrikmakait) [inline distributed/tests/test_spans.py:394]: OOT: It's mildly scary that this is even legitimate behavior.

**Review by hendrikmakait — APPROVED** [review]
Thanks, @crusaderky!

**Review by hendrikmakait — APPROVED** [review]
Thanks, @crusaderky!

File: distributed/tests/test_spans.py (hunk 2 above)
Comment (crusaderky) [reply distributed/tests/test_spans.py:394]: To be frank I think that the whole scatter design and implementation should be levelled and redone from scratch

**Pull request merged**